src/main/java/org/jenkinsci/plugins/workflow/support/DefaultStepContext.java

```diff
@@ -68,14 +68,15 @@
         if (key == EnvVars.class) {
             Run<?,?> run = get(Run.class);
             EnvironmentAction a = run == null ? null : run.getAction(EnvironmentAction.class);
-            EnvVars customEnvironment = a != null ? a.getEnvironment() : run.getEnvironment(getExecution().getOwner().getListener());
-            return key.cast(EnvironmentExpander.getEffectiveEnvironment(customEnvironment, (EnvVars) value, get(EnvironmentExpander.class)));
+            TaskListener taskListener = getListener(true);
+            EnvVars customEnvironment = a != null ? a.getEnvironment() : run.getEnvironment(taskListener);
```

Comment:
This should be OK. See https://github.com/jenkinsci/workflow-support-plugin/commit/d5d1f46255b623587198a25f8c179c64f0b74d12: there was an issue using the per-step `TaskListener` which you also encountered via a different call stack. So now we just use the per-step listener if the step is running, and the general (build) listener otherwise.